phac-nml/irida-next
# Data Export - Create Service

## What does this PR do and why?
This PR creates the `DataExports` `CreateService`. It will accept and validate parameters required to start a `DataExport` model instance before calling the `CreateExportJob` (future PR) to do the actual export creation.

## Screenshots or screen recordings
_Screenshots are required for UI changes, and strongly recommended for all other pull requests._

## How to set up and validate locally
We will use existing seeded data to test so `bin/rails db:drop/create/migrate/seed` if needed.

1. Set `user = User.find_by(email: 'user7@email.com')`
2. Set `params_1 = {'export_type' => 'sample', 'export_parameters' => {'ids' => [Sample.find_by(name: 'Yersinia pestis/Outbreak 2022 Sample 10').id]}, 'email_notification' => true, 'name' => 'test export'}`
3. Create a new `DataExport` with `DataExports::CreateService.new(user, params_1).execute`
4. Check the `export` exists under `user.data_exports.last`. 
5. Ensure  status was automatically set to `processing`

6. Repeat the above except with `user = User.find_by(email: 'user8@email.com')` (we want user8). An export should not generate with a user of access level guest for the sample (expect access level >= Analysis).

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.
